docs/design/2018-07-19-row-format.md

```diff
@@ -0,0 +1,86 @@
+The complexity is O(logN).
+
+If the column ID can not be found in the row, we use the column default value in the schema.
+
+## Rationale
+
+The new format has better read performance but we need to consider the space efficiency.
+
+In most of the cases, column IDs are less than 255 and the total length of the values is less than 64KB, so we only compare the small row type with the old format.
+
+In the current row format, a string type column takes 1 (varint flag) + 1(colID) + 1 (bytes flag) + 1(varint len) + len(val). In the new row format, a string type column takes 1(colID) + 2(offset) + len(val), one byte less. So the string type is more space efficient and varint is more space efficient.
+
+But for the decoding performance, I choose the dynamic size of 1, 2, 4, 8 to store the integers. The space efficiency depends on the value of the integer. It is more space efficient in most of the cases and it’s less space efficient in rare cases, for example if the value is slightly greater than MaxInt32.
+
+The null column IDs array is separated from not-null column IDs, so we don’t need to store offsets to null columns. It saves a lot of space when most of the columns are `null`.
+
+And we can not omit the null column ID because if the column ID is not found in the row, we will use the default value in the schema which may not be `null`.
+
+## Compatibility
+
+We can make the new row format compatible with the old row format. Because every row starts with the same varint flag byte for the column ID, with the first byte codec version that is different from the varint flag, we can support both the new row and the old row at the same time. Later in compaction process, we can rewrite the old row to the new row. After full compaction, all rows will be stored in the new format.
+
+In order to maintain the compatibility in the old cluster, we need 2 phase upgrade. The first upgrade makes the TiDB cluster able to read the new row format, but TiDB will not write data in the new row format. When all components in the cluster are upgraded in the first phase, we can do a second upgrade and after that, TiDB will write data in the new row format.
+
+TiDB tools projects need to be upgraded to support the new row format.
+
+## Implementation
+
+I implemented the new row format benchmark that decodes an encoded row into a Chunk. As a result, the performance of decoding one integer column from a row is improved up to 77x when there are 255 columns in a row. When a row has only two columns, the performance of decoding the two columns is improved 2x.
+The prototype decodes the row into a chunk instead of Datum because it is more efficient, but the chunk execution is not implemented in TiKV yet, so we can do it in mockTiKV first. When chunk execution is implemented in TiKV, we can then implement it in TiKV. An alternative approach is to change the row format before chunk execution in TiKV, then some of the work will be discarded when chunk execution is implemented.
```

Comment:
We need to benchmark the performance of decoding the new row to a Datum row, thus we can know the performance impact of the current execution method.

Reply:
I removed the decoding to chunk part, only benchmark the cut part, which is also used for Datum row decoding, the performance is improved up to near 200x.